interop-reactiveStreams/shared/src/main/scala/scalaz/zio/interop/reactiveStreams/QueueSubscriber.scala

```diff
@@ -1,56 +1,77 @@
 package scalaz.zio.interop.reactiveStreams
 
 import org.reactivestreams.{ Subscriber, Subscription }
-import scalaz.zio.stream.Take
-import scalaz.zio.stream.Take.{ End, Fail, Value }
-import scalaz.zio.{ Queue, Runtime, UIO }
-
-private[reactiveStreams] class QueueSubscriber[A](
-  runtime: Runtime[_],
-  q: Queue[Take[Throwable, A]]
-) extends Subscriber[A] {
-
-  // all signals in reactive streams are serialized, so we don't need any synchronization
-  private var subscriptionOpt: Option[Subscription] = None
-  private var signalledDemand                       = 0
-
-  def signalDemand: UIO[Unit] =
-    subscriptionOpt.fold(UIO.unit) { subscription =>
-      q.size.flatMap {
-        case 0 if q.capacity - signalledDemand > 0 => UIO(subscription.request((q.capacity - signalledDemand).toLong))
-        case _                                     => UIO.unit
-      }
+import scalaz.zio.stream.ZStream
+import scalaz.zio.stream.ZStream.Fold
+import scalaz.zio.{ IO, Queue, Runtime, Task, UIO, ZIO }
+
+private[reactiveStreams] object QueueSubscriber {
+  def make[A](capacity: Int): ZIO[Any, Nothing, (Subscriber[A], ZStream[Any, Throwable, A])] =
+    for {
+      runtime <- UIO.runtime[Any]
+      q       <- Queue.bounded[A](capacity)
+      qs      = new QueueSubscriber[A](runtime, q)
+    } yield (qs, qs)
+}
+
+private class QueueSubscriber[A](runtime: Runtime[_], q: Queue[A])
```

Comment:
How do you feel about separating the stream from the subscriber? It makes the distinction between "side-effect land" and "purity land", and will help us be more disciplined about where side effects are permitted later on.